doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+inconvenience with it. If the password is compromised on only one of the devices
+the user uses, the attacker gains full administrative access to the user's
+account and all services associated with it. The user can then halt it only by
+changing this single account's password and reconfiguring all systems and
+applications using that password for authentication (even if only to one
+particular service) to use the new one.
+
+This enhancement therefore brings support multiple passwords for a single user
+and extends the HBAC access control to allow specifying what application-specific
+passwords can be used to log into a particular service. Only allowing multiple
+passwords for a single user itself would, if application-specific passwords are
+each used only on one device, increase convenience in the case that the user can
+no longer trust a password used on one of his devices. Namely, he can revoke or
+change only that password, without needing to revoke or change his other
+passwords, including the primary password. What is more, application-specific
+passwords will generally not be usable for account management. The use of HBAC
+with them will add another security benefit to this - the intruder's access to
+the user's account might be denied not only to the account management, but also
+to at least some of the services that it is used for. The proposed HBAC
+extension would further strenghten this benefit.
+
+Resolves: https://pagure.io/freeipa/issue/4510
+How Google does it: https://support.google.com/accounts/answer/185833?hl=en
+
+## Use cases
+
+### Use with password-only authentication
+
+The user can set multiple passwords per account. Each password can be then used
+for authenticating on only one device. Various passwords can as well be used to
+authenticate to various services or their sets. These divisions can even be
+combined, thus - in the most extreme case - having one password for
+authenticating to one specific service on one particular device. Granularity of
+such divisions will be fully up to user's discretion.
+
+### Use as a part of 2FA (OTP)
+
+This may work, as the code handling the second factor is independent from the
+code handling the password(s), but it will not be a goal for this extension nor
+will it be tested.
+
+~~The user could enter an application-specific password instead of the primary
+one, and then an OTP, allowing application-specific passwords to be used as
+a part of 2FA. Although 2FA reduces the risk of account hijacking, the benefits
+of application-specific passwords still improve security in this case as well.~~
+
+### Use as a fallback mechanism when 2FA is activated, but the client
+    application does not support it
+
+This is almost certainly not needed, as authentication is managed by FreeIPA,
+which obviously supports 2FA, and client applications only consume the result of
+the authentication. And again, 2FA is out of scope of this extension.
```

Comment:
The proposal has to define how IPA should behave when OTP is mandatory and user tries to identify with an application-specific password. Google uses app passwords to support the fallback case, e.g. I'm using an app password in Thunderbird to access Google's IMAP server.

Reply:
So it is not as I assumed, that client applications only consume the final result of the authentication - that means who authenticated, and whether he authenticated successfully, regardless of the authentication method used. If client applications are more deeply involved in authentication when using FreeIPA as well, that is a different story. Then I will address this use case.

Reply:
I have just fully realized what you wrote - this means that I should also test whether app-specific passwords work with 2FA, and if not, make them work?

Reply:
Theoretically, there is another way - enable to turn 2FA on or off selectively for each app-specific password, besides the primary password. But this would be much more complicated both to implement, and to manage.

Reply:
This is not really needed, the point of an app password is to give it to an "app", therefore it makes no sense to enable 2fa as applications do not have the ability to hold two factors "a thing you know and a thing you have" because they always need to have everything recorded and available programmatically in order to work
For this reason I think that what we need is to just have a very different password policy, that enforces something like a minimum length of 20 chars and more entropy, but not require 2fa, this will make it unfeasible for someone to use this mechanism as a workaround 2fa, but hey if they remember random 20chars passwords, kudos to them ...

Reply:
So I must, conversely, make sure that when 2FA is enabled, it works only for the primary password, but not for any of the app-specific passwords, as that would cause too much trouble, right?

Reply:
Yes, that's it.

- Enable OTP as mandatory authentication method (``ipa config-mod --user-auth-type=otp``)
- Assign an OTP token to the user
- Check that the user must enter primary password + OTP to log in (the primary password is not sufficient)
- Check that the application password works without OTP

Reply:
All right, that makes sense. Thank you for clarification.

Reply:
One more question: before our discussions, I also considered app-specific passwords for RADIUS and Hardened Password authentication mechanisms, as they both also use a password for authentication. These authentication mechanisms are distinct from plain password, but as far as I know, they do not constitute 2FA. Should I leave these authentication mechanisms unaddressed - regardless of whether they will work with app-specific passwords or not - or should I also make sure that they are not used with app-specific passwords?

Reply:
You are mixing up pre-authentication mechanisms and password-based authentication. Hardened is an authentication indicator in Kerberos tickets we assign to any ticket that was obtained using pre-authentication mechanism that prevents exposure of an exchange between Kerberos client and KDC. More specifically, right now there are two of those: if exchange is utilizing SPAKE pre-authentication scheme (draft-ietf-kitten-krb-spake-preauth) or if exchange is wrapped into some existing Kerberos ticket with the help of a FAST tunneling (RFC 6113). The latter is overridden for OTP or RADIUS pre-authentication mechanism (RFC 6560) because it requires use of FAST tunnel.

The use of application-specific passwords is predominantly for non-Kerberos environments.

Reply:
All right, so Hardened password is another thing. Therefore my question applies only to RADIUS. As for your last sentence, we are going to discuss whether to make app-specific passwords availability via Kerberos optional, as noted [above](https://github.com/freeipa/freeipa/pull/4061#discussion_r400077927).

Reply:
RADIUS is all about outsourcing the actual authentication and password storage to a different server, either in RADIUS server itself or through the RADIUS server to a different place. It makes no sense to have multiple app passwords themselves stored in FreeIPA for RADIUS operation.

As for being able to authenticate with app-specific password in Kerberos, this doesn't need any specific setup because as long as there is a kerberos key for that principal (`username/app`), it will be tried by the KDC. So practically, it is either hardened already (with SPAKE on a newer systems) or could be hardened with anonymous PKINIT (on almost every system since 2010 or so).

Reply:
All right, so RADIUS is irrelevant for us. As for what you wrote about in the second paragraph - firstly, yes, I got it that we do not need to extend Hardened auth. mechanism as well. Secondly, do you mean that if we will use subaccounts with two-component principals, then an extension of ipa-kdb will not be necessary?

Reply:
It depends on what we want to define for the users as a principal in TGT. KDC sets it to the primary one in case you are using aliases. However, these principals aren't aliases, they would be separate ones (`username/app`) not really associated with the original `username` principal in KDC terms. If we want it to be the original `username` instead of `username/app`, then we'll need to modify our code.